modules/queue/workerpool.go

```diff
@@ -290,13 +326,65 @@ func (p *WorkerPool) Wait() {
 	p.cond.Wait()
 }
 
+// IsPaused returns if the pool is paused
+func (p *WorkerPool) IsPaused() bool {
```

Comment:
The meaning of `IsPaused` is quite strange. I thought it should be: if no resume, then `IsPaused` should always return true.

Actually, IsPaused is a one-time call, after the channel reading, it always return false.

Reply:
This function works exactly as described. If the pool is Paused it will always return `true` and if it is Resumed `false`. 

---

No data is ever pushed to these `p.paused` or `p.resumed`.

Therefore the only time that `<-p.paused` returns is if `p.paused` is closed, i.e. the pool is paused.

This follows the same behaviour as the `<-ctx.Done()` pattern. However, the slight complexity here is that because being Paused and Resumed are not one time things we have to deal with the fact that the paused and resumed channels may be changed - so we have to lock to check the most recent paused channel. (There are pure channel techniques but they're horrendously complicated.)

Thus:

```go
// IsPaused returns if the pool is paused
func (p *WorkerPool) IsPaused() bool {
	p.lock.Lock()
	defer p.lock.Unlock()
	select {
	case <-p.paused:
		return true
	default:
		return false
	}
}
```

* When the pool is paused `p.paused` is closed - `case <-p.paused` returns, so this returns true
* When the pool is resumed `p.paused` is open - `case <-p.paused` blocks, so default is selected and this returns false.

Although strange if you've never seen it before it's actually quite idiomatic go to use `chan struct{}`s like this - and combining them like this means you have a bool flag and blocking flag.

Now, yes we could replace the select {...} with a bool flag and but we would also have to remember when we close the channel to update the flag too but this just works.